### DISCUSSION TITLE
Stubs best practice in E2E testing

### DISCUSSION THREAD

**Discussion by Dannyisadog** [cypress-io/cypress#30470] [Questions and Help]
Hi all,  I have a question about stub a function

related issue: https://github.com/cypress-io/cypress/issues/26946

I have a page and it probably has some parent Contexts.

Some general functions will called in Context Provider and I want to stub it, how to solve it or is there any best best practice base on the scenario, thanks.

`src/utility`
```
const myFunction = () => '123';

export default {
  myFunction,
};

```

`testing.tsx`
```
import u from 'src/utility';

const TestingPage = () => (
  // parent component will call u.myFunction()
  <div>Testing</div>
);
export default TestingPage;

```

`e2eTesting.cy.ts`
```
import u from 'src/utility';

describe('Stub testing', () => {
  beforeEach(() => {
    cy.stub(u, 'myConsole').returns('456');
  });

  it('should get mocked data correctly', () => {
    cy.visit('/testing');
    // console get 123 instead of 456
  });
});

```

**Comment by portyu9** [comment]
Hi @Dannyisadog 

There are two separate issues in the sample.

First, the example stubs `myConsole` but the application calls `myFunction`.

More importantly, even if that is corrected, this pattern will not stub the application's module:

```js
import u from 'src/utility'
cy.stub(u, 'myFunction')
```

The Cypress spec and the application are separate bundles/runtime module graphs. `cy.stub()` only replaces the method on the exact object reference you pass to it; it does not rewrite the module instance already imported by the AUT.

For E2E tests, stub a stable boundary instead:

- HTTP/API behavior: `cy.intercept()`
- a dependency deliberately exposed on `window`: stub that exact `window` property in `cy.visit({ onBeforeLoad })`
- internal module functions: prefer dependency injection and test that module/component at the unit/component level

If the application continues to call `import u from 'src/utility'`, assigning or stubbing a different copy from the Cypress spec will not affect it.

Please mark this answer as accepted if it helped, thank you.
